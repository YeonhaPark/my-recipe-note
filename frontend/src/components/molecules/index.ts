import Input from './input';
import Ingredient from './ingredient';
import IngredientType from './ingredient.type';
import Tags from './tags';

export { Input, Ingredient, Tags };
export type { IngredientType };
