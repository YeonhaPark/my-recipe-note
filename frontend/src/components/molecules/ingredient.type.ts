type IngredientType = {
  id: number;
  isChecked: boolean;
  name: string;
};

export default IngredientType;
