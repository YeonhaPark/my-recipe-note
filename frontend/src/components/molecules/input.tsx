/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx, css } from '@emotion/react';
import { Input as MInput, InputLabel } from '@material-ui/core';
import { RefObject } from 'react';

interface Props {
  myRef: React.Ref<string>;
  labelName: string;
  type: string;
  placeholder: string;
  required?: boolean;
}
export default function Input({
  myRef,
  labelName,
  type,
  placeholder,
  required,
}: Props) {
  return (
    <div
      css={css`
        margin-bottom: 1.5rem;
      `}
    >
      <InputLabel htmlFor={labelName}>{labelName}</InputLabel>
      <MInput
        inputRef={myRef}
        type={type}
        id={labelName}
        fullWidth
        placeholder={placeholder}
        required={required}
      />
    </div>
  );
}

Input.defaultProps = {
  required: false,
};
