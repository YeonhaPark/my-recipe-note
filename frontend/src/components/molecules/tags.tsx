/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx, css } from '@emotion/react';
import { Chip, ChipColor } from '../atoms';
import { Autocomplete, TextField } from '@material-ui/core';
import { makeStyles } from '@material-ui/styles';
import { useState, useEffect } from 'react';
import { TagType } from '../../api/types';

const tagStyle = css`
  font-size: 0.75rem;
  flex: 0 1 1.5rem;
`;
const useStyles = makeStyles({
  root: {
    padding: '0.25rem 0',
  },
});

interface Props {
  tags: TagType[];
  setTags: React.Dispatch<React.SetStateAction<TagType[]>>;
}
const colors = [
  'primary',
  'secondary',
  'success',
  'warning',
  'default',
  'error',
  'info',
] as unknown as ChipColor;
export default function Tags({ tags, setTags }: Props) {
  const classes = useStyles();

  const [currentTagValue, setCurrentTagValue] = useState('');
  const [tagLabels, setTagLabels] = useState<string[]>(['']);

  // const addTag = (e: React.KeyboardEvent<HTMLInputElement>) => {
  //   if (e.key === 'Enter') {
  //     console.log('한번클릭했어여');
  //     const target = e.target as HTMLInputElement;
  //     const labelChipArr = tags.map((tag) => tag.label);
  //     if (!labelChipArr.includes(target.value)) {
  //       const randomColorIdx = Math.floor(Math.random() * colors.length);
  //       const randomColor: ChipColor = colors[randomColorIdx] as ChipColor;
  //       setTags([...tags, { label: target.value, color: randomColor }]);
  //     }
  //   }
  // }; 삭제해도될듯?

  const handleChange = (e: object, value: string[], reason: string) => {
    if (reason === 'clear') {
      setTags([]);
    }
    setTagLabels(value);
  };

  useEffect(() => {
    setTagLabels(tags.map((tag) => tag.label));
  }, [tags]);

  useEffect(() => {
    const event = new Event('change', {
      bubbles: true,
    });
    const node = document.getElementById('tags-standard');
    node && node.dispatchEvent(event);
  }, []);

  return (
    <div css={tagStyle}>
      <Autocomplete
        className={classes.root}
        onChange={handleChange}
        multiple
        id="tags-standard"
        options={tags.map((option) => option.label)}
        freeSolo
        value={tagLabels}
        renderTags={(value, getTagProps) => {
          const newTags = value.map((val: string, index: number) => {
            return {
              label: val,
              color: colors[index] as ChipColor,
            };
          });
          const chips = newTags.map((tag: TagType, index: number) => (
            <Chip
              {...getTagProps({ index })}
              key={tag.label}
              color={tag.color}
              variant="filled"
              label={tag.label}
            />
          ));
          return (
            <div
              css={css`
                margin: 0.25rem auto;
              `}
            >
              {chips}
            </div>
          );
        }}
        renderInput={(params) => {
          return (
            <TextField
              {...params}
              onChange={() => setCurrentTagValue}
              value={currentTagValue}
              variant="standard"
              label="Add tags"
            />
          );
        }}
      />
    </div>
  );
}
