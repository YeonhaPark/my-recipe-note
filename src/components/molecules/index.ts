import Input from './input';
import Ingredient from './ingredient';

export { Input, Ingredient };
