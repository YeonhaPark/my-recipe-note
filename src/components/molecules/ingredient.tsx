/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx, css } from '@emotion/react';
import React from 'react';
import { Checkbox } from '../atoms';

const checkboxStyle = css`
  display: flex;
  align-items: center;
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
  input {
    border: 0;
  }
`;

interface Props {
  isFocused: boolean;
  checked: boolean;
  onValueChange: (
    type: string,
    id: number,
    e: React.ChangeEvent<HTMLInputElement>,
  ) => void;
  onEnter: (e: React.KeyboardEvent<HTMLInputElement>) => void;
  value: string;
  id: number;
}
export default function Ingredient({
  isFocused,
  checked,
  onEnter,
  onValueChange,
  value,
  id,
}: Props) {
  return (
    <div css={checkboxStyle}>
      <label htmlFor="ingredient">
        <Checkbox
          id={id}
          checked={checked}
          onCheckboxChange={onValueChange}
          name={`checkbox_${id}`}
          color="primary"
        />
      </label>
      <input
        autoFocus={isFocused}
        id="ingredient"
        type="text"
        value={value}
        onKeyDown={onEnter}
        onChange={(e) => onValueChange('text', id, e)}
      />
    </div>
  );
}
