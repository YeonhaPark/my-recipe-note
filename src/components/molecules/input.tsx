/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx, css } from '@emotion/react';
import { Input as MInput, InputLabel } from '@material-ui/core';

interface Props {
  labelName: string;
  type: string;
  placeholder: string;
  required?: boolean;
}
export default function Input({
  labelName,
  type,
  placeholder,
  required,
}: Props) {
  return (
    <div
      css={css`
        margin-bottom: 1.5rem;
      `}
    >
      <InputLabel htmlFor={labelName}>{labelName}</InputLabel>
      <MInput
        type={type}
        id={labelName}
        fullWidth
        placeholder={placeholder}
        required={required}
      />
    </div>
  );
}

Input.defaultProps = {
  required: false,
};
