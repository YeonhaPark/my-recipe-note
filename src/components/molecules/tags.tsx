/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx, css } from '@emotion/react';
import { Chip, ChipType, ChipColor } from '../atoms';
import { Autocomplete, TextField } from '@material-ui/core';
import { makeStyles } from '@material-ui/styles';
import React from 'react';

const tagStyle = css`
  font-size: 0.75rem;
  flex: 0 1 1.5rem;
`;
const useStyles = makeStyles({
  root: {
    padding: '0.25rem 0',
  },
});

interface Props {
  setTag: React.Dispatch<React.SetStateAction<string>>;
  tag: string;
  tagList: ChipType[];
  setTagList: React.Dispatch<React.SetStateAction<ChipType[]>>;
}
const colors = [
  'primary',
  'secondary',
  'success',
  'warning',
  'default',
  'error',
  'info',
] as unknown as ChipColor;
export default function Tags({ setTag, tag, tagList, setTagList }: Props) {
  const classes = useStyles();

  const addTag = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      const target = e.target as HTMLInputElement;
      const labelChipArr = tagList.map((tag) => tag.label);
      if (!labelChipArr.includes(target.value)) {
        const randomColorIdx = Math.floor(Math.random() * colors.length);
        const randomColor: ChipColor = colors[randomColorIdx] as ChipColor;
        setTagList([...tagList, { label: target.value, color: randomColor }]);
      } else {
      }
    }
  };

  const handleChange = (event: object, value: string[], reason: string) => {
    if (reason === 'clear') {
      setTagList([]);
    }
  };
  return (
    <div css={tagStyle}>
      <Autocomplete
        className={classes.root}
        onChange={handleChange}
        multiple
        id="tags-standard"
        options={tagList.map((option) => option.label)}
        defaultValue={undefined}
        freeSolo
        renderTags={(value, getTagProps) => {
          const newValues = value.map((val: string, index: number) => {
            return {
              label: val,
              color: colors[index] as ChipColor,
            };
          });
          const chips = newValues.map((tag: ChipType, index: number) => (
            <Chip
              color={tag.color}
              variant="filled"
              label={tag.label}
              {...getTagProps({ index })}
            />
          ));
          return (
            <div
              css={css`
                margin: 0.25rem auto;
              `}
            >
              {chips}
            </div>
          );
        }}
        renderInput={(params) => (
          <TextField
            {...params}
            onChange={() => setTag}
            value={tag}
            onKeyDown={addTag}
            variant="standard"
            label="Add tags"
          />
        )}
      />
    </div>
  );
}
