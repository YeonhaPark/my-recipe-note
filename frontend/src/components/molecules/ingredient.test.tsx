import { mount } from 'enzyme';
import { Ingredient } from '../molecules';
import renderer from 'react-test-renderer';

const IngredientProps = {
  id: 1,
  checked: true,
  onValueChange: jest.fn(),
  onCheckboxChange: jest.fn(),
  onListChange: jest.fn(),
  value: 'bread',
  ingredientId: 1,
};

const renderIngredient = () => mount(<Ingredient {...IngredientProps} />);

describe('Ingredient', () => {
  it('renders Ingredient', () => {
    renderIngredient();
  });

  it('renders exact prop provided', () => {
    const ingredient = renderIngredient();

    expect(ingredient.prop('value')).toEqual('bread');
  });

  it('contains input tag as child', () => {
    const ingredient = renderIngredient();
    expect(ingredient.find('input#input-1')).toHaveLength(1);
  });
});
