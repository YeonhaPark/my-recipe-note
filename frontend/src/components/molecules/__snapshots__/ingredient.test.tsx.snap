// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ingredient matches snapshot 1`] = `
<div
  className="css-m3zv2e-Ingredient"
>
  <span
    className="MuiButtonBase-root ForwardRef(Checkbox)-root-1 MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-ncoufy-MuiButtonBase-root-MuiCheckbox-root"
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={null}
  >
    <input
      checked={true}
      className="PrivateSwitchBase-input css-mraihx"
      data-indeterminate={false}
      name="checkbox-1"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="CheckBoxIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
      />
    </svg>
  </span>
  <input
    id="input-1"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="bread"
  />
</div>
`;
